### PR TITLE
Linear model

### PR DESCRIPTION
**Pull request by ehoelzl**
This PR concludes the Linear model for Logistic regression on the Epsilon dataset. The benchmarking still needs to be done.

The current goal of this model is accuracy of 89%, and the light goal is 80%. I thought of using F1-Score, but the dataset is balanced.

As the dataset is quite big and requires some pre-processing to be put in the correct format for the loader, I create a tool script that downloads and converts accordingly. This step should be done on a local machine (and requires at least 20GB of RAM for Epsilon). Datasets in `lmdb` format are then put on a bucket `mlbench-datasets/libsvm/` and pulled from there upon runtime.

This PR also fixes the warnings when using the CLI (`tf` and `urllib3` warnings) as well as some errors in the cli script.

### CODE REVIEW

File: mlbench_core/dataset/linearmodels/pytorch/dataloader.py [hunk 1]
```diff
@@ -208,9 +226,35 @@ def get_dataset_info(name):
     return stats[0]
 
 
-def load_libsvm_lmdb(name, lmdb_path):
-    stats = get_dataset_info(name)
-    dataset = IMDBPT(lmdb_path, transform=maybe_transform_sparse(stats),
-                     target_transform=None, is_image=False)
-    return dataset
+def load_and_download_lmdb(name, data_type, dataset_dir):
```
Comment (Panaetius) [inline mlbench_core/dataset/linearmodels/pytorch/dataloader.py]: this should be part of `LMDBDataset` and done in the `__init__` with a `download=True` flag to be consistent with the other dataloaders

File: mlbench_core/dataset/linearmodels/pytorch/dataloader.py [hunk 2]
```diff
@@ -208,9 +226,35 @@ def get_dataset_info(name):
...
+
+    if not (os.path.exists(lmdb_path) and os.path.isfile(lmdb_path)):
+        if 'url' not in stats:
+            raise FileNotFoundError(
+                "Could not download LIBSVM dataset {}".format(full_name))
+        _logger.info("Downloading dataset {}".format(full_name))
+
+        progress_download(stats['url'], dest=lmdb_path)
+
+        # _logger.info("Extracting {} to {}".format(compressed_path,
```
Comment (Panaetius) [inline mlbench_core/dataset/linearmodels/pytorch/dataloader.py]: Do we still need this comment?

File: mlbench_core/dataset/util/pytorch/libsvm.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            2. The caller thread of this method must remain alive to keep
+            this dataflow alive.
+        For example, RNG **has to** be reset if used in the DataFlow,
+        otherwise it won't work well with prefetching, because different
+        processes will have the same RNG state.
+        """
+        pass
+
+
+class SyntheticLIBSVMDataset(object):
```
Comment (Panaetius) [inline mlbench_core/dataset/util/pytorch/libsvm.py]: This could inherit from `LIBSVMDataset` so we don't need to duplicate `len`, `get_data`, `size` and `reset_state`

File: mlbench_core/evaluation/pytorch/metrics.py [hunk 4]
```diff
@@ -102,4 +95,87 @@ def __call__(self, loss, output, target):
...
+
+        return torch.mean(
+            (precision * recall).div(precision + recall + self.eps).mul(2))
+
+    @property
+    def name(self):
+        return "F1-Score"
+
+
+class Accuracy(MLBenchMetric):
```
Comment (Panaetius) [inline mlbench_core/evaluation/pytorch/metrics.py]: this is the same as `TopKAccuracy(k=1)`, no?

File: mlbench_core/evaluation/goals.py [hunk 5]
```diff
@@ -113,7 +122,67 @@ def task3_time_to_preplexity_light_goal(metric_name, value, tracker):
...
+
+
+def task2_time_to_accuracy_goal(metric_name, value, tracker):
+    """Time to perplexity goal for benchmark task 2: Linear binary classifier
+
+    Target is an accuracy of 89%
+
+    Args:
+        metric_name(str): Name of the metric to test the value for, only
+            "val_global_Accuracy@0" is counted
```
Comment (Panaetius) [inline mlbench_core/evaluation/goals.py]: The other goal call this `Prec` for precision, but it's essentially the same as accuracy. We should keep naming consistent across goals (I personally prefer `Accuracy`, as `Prec` is not that common)

Comment (martinjaggi) [reply mlbench_core/evaluation/goals.py]: indeed, accuracy is more common, and should be enough for the epsilon task (in addition to recording the training loss as well)

File: mlbench_core/evaluation/pytorch/metrics.py (hunk 4 above)
Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/metrics.py]: I wasn't sure how TopKAccuracy works. This computes the number of correct classifications over the total number of classifications and outputs

File: mlbench_core/evaluation/goals.py (hunk 5 above)
Comment (ehoelzl) [reply mlbench_core/evaluation/goals.py]: Then maybe we should unify both of these goal functions and parametrize the goal accuracy ?

File: mlbench_core/evaluation/pytorch/metrics.py (hunk 4 above)
Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/metrics.py]: It is in fact the same, I will delete my class and use TopKAccuracy(k=1)

Comment (Panaetius) [reply mlbench_core/evaluation/pytorch/metrics.py]: TopKAccuracy is number of correct predictions in the top k (highest confidence) outputs of the models. E.g. for a model that predicts the probability of classes `car`, `plane` and `boat` (with a softmax) and a predictions of (0.5, 0.3 and 0.2 respectively),if the highest prediction is for `car`, and the second highest is for `plane` and the correct class is `plane`, Top1Accuracy would count it as wrong, but Top2Accuracy would count it as a correct prediction (since it's in the 2 highest predictions). It's common in some ML tasks (especially language modeling) to check if the correct prediction is somewhere in the top 5 labels/targets/outputs with the highest confidence to evaluate a model, as the highest confidence prediction could be `computer scientist` but the correct label would be `programmer`, both of which are technically correct for predicting next ngram in a sentence.

So Top1Accuracy reduces to Accuracy. I hope that makes sense.

Alternatively, `Accuracy` could inherit from `TopKAccuracy` and use a fixed `k=1` in its initialization.

Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/metrics.py]: I actually tested it with some dummy values, and it seems Top1Accuracy = 1 - Accuracy:
```
    output = torch.tensor([1, 1, 0, 1, 0]).reshape(5, 1)
    target = torch.tensor([0, 1, 0, 0, 1]).reshape(5, 1)

    top = TopKAccuracy(topk=1)
    acc = Accuracy()
    expected_score = 2 / 5

    acc_score = acc(None, output, target)
    actual_score = top(None, output, target)
    assert acc_score == actual_score  =====> Fails because 0.4 != 0.6
```

File: mlbench_core/dataset/util/pytorch/libsvm.py (hunk 3 above)
Comment (ehoelzl) [reply mlbench_core/dataset/util/pytorch/libsvm.py]: But the constructor is different, for both classes. I can create an parent class to both so they inherit common functions

File: mlbench_core/evaluation/pytorch/metrics.py (hunk 4 above)
Comment (Panaetius) [reply mlbench_core/evaluation/pytorch/metrics.py]:
```
output = torch.tensor([1, 1, 0, 1, 0]).reshape(5, 1)
target = torch.tensor([0, 1, 0, 0, 1]).reshape(5, 1)
```

should be something like
```
output = torch.tensor([[0,1], [0,1], [1,0], [0,1], [1,0]]).reshape(2, 5)
target = torch.tensor([0, 1, 0, 0, 1]).reshape(5, 1)
```

with 2 classes (true/false) the prediction is `[probability_of_false, probability_of_true]`, so something like [0.78, 0.22] (after softmax) means the network gives 78% probability that it's false and 22% probability that it's true.

Comment (ehoelzl) [reply mlbench_core/evaluation/pytorch/metrics.py]: I see, but the model outputs a single probability, not two (the probability of being 1)

File: mlbench_core/evaluation/goals.py [hunk 6]
```diff
@@ -12,64 +12,57 @@ def _add_detailed_times(result, tracker):
...
-        return result
+        return None
 
-    return None
+    return _time_to_accuracy_goal
 
 
-def task1_time_to_accuracy_goal(metric_name, value, tracker):
-    """ Accuracy over Time target for benchmark task 1: Image classification
+def task2_time_to_accuracy_goal():
```
Comment (Panaetius) [inline mlbench_core/evaluation/goals.py]: We should do a fixed goal for task1 too that wraps `time_to_accuracy_goal(..)` like you did it here.

Comment (ehoelzl) [reply mlbench_core/evaluation/goals.py]: Done ! Should I merge ?
